### PR TITLE
fix(developer): handle second parameter of index correctly in kmcmplib compiler

### PR DESCRIPTION
**Pull request by markcsinclair**
Fix the issues with the handling of a non-integer second parameter (offset) to `index` in the `kmcmplib` compiler - see issue [#11814](https://github.com/keymanapp/keyman/issues/11814)

Tests for `index` are included in `gtest-compiler-test.cpp`

Partially fixes: #11814 (another PR will be needed for the second bug)

@keymanapp-test-bot skip

### CODE REVIEW

**Comment by ermshiperete** [comment]
Note the failing test on TC though:

```
...
58/58 gtest-compiler-test                     FAIL            0.27s   (exit status 3221226356 or signal 3221226228 SIGinvalid)
15:43:41   >>> MALLOC_PERTURB_=25 C:\BuildAgent\work\7ac43416c45637e9\keyman\developer\src\kmcmplib\build\x86\release\tests\gtest-compiler-test.exe
15:43:41   ------------------------------------- 8< -------------------------------------
15:43:41   Running main() from ../../../subprojects/googletest-1.14.0/googletest/src/gtest_main.cc
15:43:41   
15:43:41   [==========] Running 15 tests from 1 test suite.
15:43:41   
15:43:41   [----------] Global test environment set-up.
15:43:41   
15:43:41   [----------] 15 tests from CompilerTest
15:43:41   
15:43:41   [ RUN      ] CompilerTest.strtowstr_test
15:43:41   
15:43:41   [       OK ] CompilerTest.strtowstr_test (0 ms)
15:43:41   
15:43:41   [ RUN      ] CompilerTest.wstrtostr_test
15:43:41   
15:43:41   [       OK ] CompilerTest.wstrtostr_test (0 ms)
15:43:41   
15:43:41   [ RUN      ] CompilerTest.AddCompileWarning_test
15:43:41   
15:43:41   [       OK ] CompilerTest.AddCompileWarning_test (0 ms)
15:43:41   
15:43:41   [ RUN      ] CompilerTest.AddCompileError_test
15:43:41   
15:43:41   [       OK ] CompilerTest.AddCompileError_test (0 ms)
15:43:41   
15:43:41   [ RUN      ] CompilerTest.ProcessBeginLine_test
15:43:41   
15:43:41   [       OK ] CompilerTest.ProcessBeginLine_test (0 ms)
15:43:41   
15:43:41   [ RUN      ] CompilerTest.ValidateMatchNomatchOutput_test
15:43:41   
15:43:41   [       OK ] CompilerTest.ValidateMatchNomatchOutput_test (0 ms)
15:43:41   
15:43:41   [ RUN      ] CompilerTest.IsValidKeyboardVersion_test
15:43:41   
15:43:41   [       OK ] CompilerTest.IsValidKeyboardVersion_test (0 ms)
15:43:41   
15:43:41   [ RUN      ] CompilerTest.GetXStringImpl_test
15:43:41   
15:43:41   [       OK ] CompilerTest.GetXStringImpl_test (0 ms)
15:43:41   
15:43:41   [ RUN      ] CompilerTest.GetXStringImpl_type_xd_test
15:43:41   
15:43:41   [       OK ] CompilerTest.GetXStringImpl_type_xd_test (0 ms)
15:43:41   
15:43:41   [ RUN      ] CompilerTest.GetXStringImpl_type_double_quote_test
15:43:41   
15:43:41   [       OK ] CompilerTest.GetXStringImpl_type_double_quote_test (0 ms)
15:43:41   
15:43:41   [ RUN      ] CompilerTest.GetXStringImpl_type_single_quote_test
15:43:41   
15:43:41   [       OK ] CompilerTest.GetXStringImpl_type_single_quote_test (0 ms)
15:43:41   
15:43:41   [ RUN      ] CompilerTest.GetXStringImpl_type_i_test
15:43:41   
15:43:41   ------------------------------------------------------------------------------
15:43:41   
15:43:41   
15:43:41   Summary of Failures:
15:43:41   
15:43:41   58/58 gtest-compiler-test                     FAIL            0.27s   (exit status 3221226356 or signal 3221226228 SIGinvalid)
```

**Comment by markcsinclair** [comment]
Trouble with deletes *sigh*

File: developer/src/kmcmplib/src/Compiler.cpp [hunk 1]
```diff
@@ -3362,6 +3363,18 @@ int atoiW(PKMX_WCHAR p)
   return i;
 }
 
+bool isuiw(PKMX_WCHAR p) {
```
Comment (mcdurdin) [inline developer/src/kmcmplib/src/Compiler.cpp:3366]: I don't really know what this means -- can we use a longer name e.g. is_integer_wstring?

File: developer/src/kmcmplib/src/Compiler.cpp [hunk 2]
```diff
@@ -3362,6 +3363,18 @@ int atoiW(PKMX_WCHAR p)
...
+  if (!p || !*p)
+    return false;
+  PKMX_STR q = wstrtostr(p);
```
Comment (mcdurdin) [inline developer/src/kmcmplib/src/Compiler.cpp:3369]: This is leaking memory because `q` is not deleted

File: developer/src/kmcmplib/src/kmx_u16.cpp [hunk 3]
```diff
@@ -259,7 +259,7 @@ KMX_WCHAR * u16tok(KMX_WCHAR* p, const KMX_WCHAR* delim, KMX_WCHAR** ctx) {
 	if (*q) {
 		*q = 0;
 		q++;
-		while (u16chr(delim, *q)) q++;
+		while (*q && u16chr(delim, *q)) q++;
```
Comment (mcdurdin) [inline developer/src/kmcmplib/src/kmx_u16.cpp:262]: This is a separate bug: it needs to be tracked in a separate PR. I think this is important for regression management.

Two further opportunities here:
1. Audit other uses of `u16chr` for the same bug elsewhere in the compiler.
2. Add a unit test for `u16tok`, looking at existing usage for current patterns.

File: developer/src/kmcmplib/tests/gtest-compiler-test.cpp [hunk 4]
```diff
@@ -401,6 +402,76 @@ TEST_F(CompilerTest, GetXStringImpl_type_single_quote_test) {
...
+    u16cpy(str, u"index(");
+    EXPECT_EQ(CERR_InvalidIndex, GetXStringImpl(tstr, &fileKeyboard, str, u"", output, 80, 0, &newp, FALSE));
+
+    // index, empty delimiters => empty string
+    u16cpy(str, u"index()");
+    EXPECT_EQ(CERR_InvalidIndex, GetXStringImpl(tstr, &fileKeyboard, str, u"", output, 80, 0, &newp, FALSE));
+
+    // index, no comma or space
+    u16cpy(str, u"index(b)");
+    EXPECT_EQ(CERR_InvalidIndex, GetXStringImpl(tstr, &fileKeyboard, str, u"", output, 80, 0, &newp, FALSE));
```
Comment (mcdurdin) [inline developer/src/kmcmplib/tests/gtest-compiler-test.cpp:432]: Perhaps additional tests for whitespace inside parentheses?

File: developer/src/kmcmplib/tests/gtest-compiler-test.cpp [hunk 5]
```diff
@@ -401,6 +402,76 @@ TEST_F(CompilerTest, GetXStringImpl_type_single_quote_test) {
...
+    // index, comma, valid
+    fileKeyboard.cxStoreArray = 3u;
+    fileKeyboard.dpStoreArray = file_store;
+    u16cpy(str, u"index(b,4)");
+    EXPECT_EQ(CERR_None, GetXStringImpl(tstr, &fileKeyboard, str, u"", output, 80, 0, &newp, FALSE));
+    const KMX_WCHAR tstr_index_comma_valid[] = { UC_SENTINEL, CODE_INDEX, 2, 4, 0 };
+    EXPECT_EQ(0, u16cmp(tstr_index_comma_valid, tstr));
+
+    // index, space, valid
+    u16cpy(str, u"index(b 4)");
```
Comment (mcdurdin) [inline developer/src/kmcmplib/tests/gtest-compiler-test.cpp:447]: This should not be valid. See https://help.keyman.com/developer/language/reference/index. I recognize we do not have a formal grammar for .kmn, but comma is expected between parameters.

File: developer/src/kmcmplib/tests/gtest-compiler-test.cpp [hunk 6]
```diff
@@ -401,6 +402,76 @@ TEST_F(CompilerTest, GetXStringImpl_type_single_quote_test) {
...
+    EXPECT_EQ(0, u16cmp(tstr_index_two_digit_valid, tstr));
+
+    // index, comma, non-digit parameter, CERR_InvalidIndex
+    u16cpy(str, u"index(b,g)");
+    fileKeyboard.cxStoreArray = 3u;
+    fileKeyboard.dpStoreArray = file_store;
+    EXPECT_EQ(CERR_InvalidIndex, GetXStringImpl(tstr, &fileKeyboard, str, u"", output, 80, 0, &newp, FALSE));
+
+    // index, comma, no parameter, CERR_InvalidIndex
+    u16cpy(str, u"index(b,)");
```
Comment (mcdurdin) [inline developer/src/kmcmplib/tests/gtest-compiler-test.cpp:469]: Add whitespace versions, e.g. `index( b, )`, `index( b ,)`?

File: developer/src/kmcmplib/tests/gtest-compiler-test.cpp [hunk 7]
```diff
@@ -442,6 +513,16 @@ TEST_F(CompilerTest, GetRHS_test) {
...
 // int atoiW(PKMX_WCHAR p)
 // KMX_DWORD kmcmp::CheckUTF16(int n)
+
+TEST_F(CompilerTest, isuiw_test) {
+    EXPECT_FALSE(isuiw(nullptr));
+    EXPECT_FALSE(isuiw((PKMX_WCHAR)u""));
+    EXPECT_FALSE(isuiw((PKMX_WCHAR)u"a"));
+    EXPECT_FALSE(isuiw((PKMX_WCHAR)u"-1"));
+    EXPECT_TRUE(isuiw((PKMX_WCHAR)u"1"));
+    EXPECT_TRUE(isuiw((PKMX_WCHAR)u"42"));
```
Comment (mcdurdin) [inline developer/src/kmcmplib/tests/gtest-compiler-test.cpp:523]:
```suggestion
    EXPECT_TRUE(isuiw((PKMX_WCHAR)u"42"));
    EXPECT_FALSE(isuiw((PKMX_WCHAR)u"2147483648")); // INT32_MAX + 1
    EXPECT_FALSE(isuiw((PKMX_WCHAR)u"9999999999")); // > INT32_MAX
```

Thinking signed integer values here... if we are going to protect against invalid data then these are going to be important.

File: developer/src/kmcmplib/tests/gtest-compiler-test.cpp [hunk 8]
```diff
@@ -18,6 +18,7 @@ KMX_DWORD GetXStringImpl(PKMX_WCHAR tstr, PFILE_KEYBOARD fk, PKMX_WCHAR str, KMX
   PKMX_WCHAR output, int max, int offset, PKMX_WCHAR *newp, int isUnicode
 );
 KMX_DWORD GetRHS(PFILE_KEYBOARD fk, PKMX_WCHAR p, PKMX_WCHAR buf, int bufsize, int offset, int IsUnicode);
+bool isuiw(PKMX_WCHAR p);
```
Comment (mcdurdin) [inline developer/src/kmcmplib/tests/gtest-compiler-test.cpp:21]:
```suggestion
bool is_integer_wstring(const PKMX_WCHAR p);
```

**Review by mcdurdin — CHANGES_REQUESTED** [review]
Some changes requested please!

In answer to your questions:

1. `isuiw` is not clear to me. Can we use a longer name, e.g. `is_integer_wstring`?
2. `CERR_InvalidIndex` is a little bit broad but I think it is okay; adding extra error codes does not add a lot of value.

**Comment by mcdurdin** [comment]
The commit history for this PR seems a little busy -- why is that?

File: developer/src/kmcmplib/tests/gtest-compiler-test.cpp (hunk 5 above)
Comment (mcdurdin) [reply developer/src/kmcmplib/tests/gtest-compiler-test.cpp:447]: Formal grammar is another development opportunity ... would be good for various other projects in the future too!

File: developer/src/kmcmplib/src/kmx_u16.cpp (hunk 3 above)
Comment (mcdurdin) [reply developer/src/kmcmplib/src/kmx_u16.cpp:262]: I recognize that this is my fault for saying 'put both of these bugs into one issue'. Sorry!

File: developer/src/kmcmplib/src/Compiler.cpp (hunk 2 above)
Comment (markcsinclair) [reply developer/src/kmcmplib/src/Compiler.cpp:3369]: Good catch ... I am spending too much time with more modern languages. Corrected.

**Comment by markcsinclair** [comment]
> The commit history for this PR seems a little busy -- why is that?

I'm not sure ... I switched to master and checked out `gtest-compiler-test.cpp` from PR [#11663](https://github.com/keymanapp/keyman/pull/11663) and cut it down to just the relevant additional test cases cf. master, but this seems to have pulled in some unexpected commits.  It may have been better to just paste the test cases from PR [#11663](https://github.com/keymanapp/keyman/pull/11663) into `gtest-compiler-test.cpp`

File: developer/src/kmcmplib/tests/gtest-compiler-test.cpp (hunk 5 above)
Comment (markcsinclair) [reply developer/src/kmcmplib/tests/gtest-compiler-test.cpp:447]: I had not found the [Keyman Keyboard Language Reference](https://help.keyman.com/developer/language/reference/), but had been working entirely from the code ... this should make life a little easier.

File: developer/src/kmcmplib/tests/gtest-compiler-test.cpp (hunk 8 above)
Comment (markcsinclair) [reply developer/src/kmcmplib/tests/gtest-compiler-test.cpp:21]: Going with camel case: `isUnsignedIntegerWstring()`

File: developer/src/kmcmplib/tests/gtest-compiler-test.cpp (hunk 7 above)
Comment (markcsinclair) [reply developer/src/kmcmplib/tests/gtest-compiler-test.cpp]: I have extended `isIntegerWstring()` to handle the full range of integer values, and added additional tests.

Comment (markcsinclair) [reply developer/src/kmcmplib/tests/gtest-compiler-test.cpp]: Also added a check for offset < 1 and supporting test cases

File: developer/src/kmcmplib/src/kmx_u16.cpp (hunk 3 above)
Comment (markcsinclair) [reply developer/src/kmcmplib/src/kmx_u16.cpp:262]: I will leave a single issue ([#11814](https://github.com/keymanapp/keyman/issues/11814)), but break out a new PR for the missing second parameter bug.

File: developer/src/kmcmplib/src/Compiler.cpp (hunk 1 above)
Comment (markcsinclair) [reply developer/src/kmcmplib/src/Compiler.cpp]: See below.

File: developer/src/kmcmplib/src/kmx_u16.cpp (hunk 3 above)
Comment (markcsinclair) [reply developer/src/kmcmplib/src/kmx_u16.cpp]: I have reversed the fix, although other checks in the index section of `GetXStringImpl()` prevent the bug from manifesting.

File: developer/src/kmcmplib/tests/gtest-compiler-test.cpp (hunk 5 above)
Comment (markcsinclair) [reply developer/src/kmcmplib/tests/gtest-compiler-test.cpp:459]: I will break out a new issue for this.

File: developer/src/kmcmplib/tests/gtest-compiler-test.cpp (hunk 4 above)
Comment (markcsinclair) [reply developer/src/kmcmplib/tests/gtest-compiler-test.cpp:432]: `index( )` (single space) also seems to suiffer from the runaway problem in `u16tok()`.

Comment (markcsinclair) [reply developer/src/kmcmplib/tests/gtest-compiler-test.cpp:432]: `index( b)` and `index(b )` added

File: developer/src/kmcmplib/tests/gtest-compiler-test.cpp (hunk 6 above)
Comment (markcsinclair) [reply developer/src/kmcmplib/tests/gtest-compiler-test.cpp:481]: `index( b,4)`, `index(b, 4)`, `index(b ,)` and `index(b, )` added

**Comment by markcsinclair** [comment]
Ready for review again

**Review by mcdurdin — APPROVED** [review]

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 18.0.66-alpha](https://keyman.com/downloads/releases/alpha/18.0.66)
